### PR TITLE
fix   Issue 11176 - array.ptr in @safe code may point past end of array

### PR DESCRIPTION
**Pull request by WalterBright**
https://issues.dlang.org/show_bug.cgi?id=11176

### CODE REVIEW

**Comment by WalterBright** [comment]
This is likely to be disruptive.

**Comment by WalterBright** [comment]
Depends on https://github.com/dlang/druntime/pull/1590

**Comment by quickfur** [comment]
LGTM. I like this, it's a neat solution to the nasty can o' worms about when/whether/how `.ptr` should/shouldn't be allowed in `@safe` code. Disruptive, yes, but it does solve the problem in a nice, self-contained way.

**Comment by JackStouffer** [comment]
Looks like there are still some std.format issues

**Comment by andralex** [comment]
I'll mark this for merging.

**Comment by andralex** [comment]
Auto-merge toggled on

**Comment by WalterBright** [comment]
This is blocked by numerous Phobos issues with `.ptr`

**Comment by WalterBright** [comment]
Auto-merge toggled off

**Comment by WalterBright** [comment]
Turned off auto-merge for the time being so the autotester doesn't waste time on this while it is blocked.

**Comment by MartinNowak** [comment]
And again, and again, and again: Stop breaking code!

> This is likely to be disruptive.

You already mention it yourself, but still do it.

> Depends on dlang/druntime#1590 and dlang/phobos#4426

As always, this is a lie. You're about to break almost 1K dub packages that all need similar PRs.
The mechanism to do this is simple and will established, instead of breaking code, you introduce a deprecation that informs people about the upcoming error, then turn it into an error a few month later when the majority of packages is fixed.

File: src/mtype.d [hunk 1]
```diff
@@ -4908,6 +4908,11 @@ extern (C++) final class TypeSArray : TypeArray
                 e.error("%s is not an expression", e.toChars());
                 return new ErrorExp();
             }
+            else if (sc.func && !sc.intypeof && sc.func.setUnsafe())
+            {
+                e.error("%s.ptr cannot be used in @safe code, use &%s[0] instead", e.toChars(), e.toChars());
```
Comment (MartinNowak) [inline src/mtype.d]: deprecated, simple as that

**Comment by MartinNowak** [comment]
This PR is also lacking communication (changelog) and a spec update.

**Comment by WalterBright** [comment]
It is breaking much more than I thought it would. Once Phobos is completely updated, I'll change it to a deprecation - but this may be tricky, as it influences the inference of @safe attributes.

As for communication, there's still the bug report for it https://issues.dlang.org/show_bug.cgi?id=11176

**Comment by schveiguy** [comment]
> there's still the bug report for it

I think this change warrants a more in-depth changelog entry.

**Comment by WalterBright** [comment]
I changed it so thing like `a.ptr == null` do not trigger the safety error, since it's only a problem if a.ptr might be dereferenced, and such uses are not dereferenced. Unfortunately, this adds a lot of lines of changes, although it should cut down dramatically on code breakage.

File: src/mtype.d [hunk 2]
```diff
@@ -2704,13 +2704,13 @@ extern (C++) abstract class Type : RootObject
                 auto dti = new DotTemplateInstanceExp(e.loc, e, Id.opDispatch, tiargs);
                 dti.ti.tempdecl = td;
                 /* opDispatch, which doesn't need IFTI,  may occur instantiate error.
-                 * It should be gagged if flag != 0.
+                 * It should be gagged if flag & 1.
                  * e.g.
                  *  tempalte opDispatch(name) if (isValid!name) { ... }
```
Comment (wilzbach) [inline src/mtype.d]: just noticed that `template`  has a typo here ;-)

**Comment by andralex** [comment]
@WalterBright guess you should close this for now?

**Comment by WalterBright** [comment]
Why close it now? At least it is finding safety issues in Phobos, which we are fixing one after the other. After that all passes, probably this PR should be adjusted so it is enabled with a switch, as it seems to require code to have a lot of fixes.

**Comment by WalterBright** [comment]
Blocked by https://github.com/dlang/dmd/pull/6097

**Comment by WalterBright** [comment]
This one is ready-to-merge now.

File: src/mtype.d [hunk 3]
```diff
@@ -4923,6 +4923,11 @@ extern (C++) final class TypeSArray : TypeArray
                 e.error("%s is not an expression", e.toChars());
                 return new ErrorExp();
             }
+            else if (!(flag & 2) && sc.func && !sc.intypeof && sc.func.setUnsafe() && global.params.safe)
```
Comment (andralex) [inline src/mtype.d]: so what's `2` again?

File: src/mtype.d [hunk 4]
```diff
@@ -5214,6 +5219,11 @@ extern (C++) final class TypeDArray : TypeArray
         }
         else if (ident == Id.ptr)
         {
+            if (!(flag & 2) && sc.func && !sc.intypeof && sc.func.setUnsafe() && global.params.safe)
```
Comment (andralex) [inline src/mtype.d]: `2`

**Comment by andralex** [comment]
Let's use a symbol for that `2` before pulling this Thanks!

File: src/mtype.d (hunk 3 above)
Comment (WalterBright) [inline src/mtype.d]: done

File: src/mtype.d (hunk 4 above)
Comment (WalterBright) [inline src/mtype.d]: done

**Comment by WalterBright** [comment]
done

**Pull request merged**